extra_data/components.py

```diff
@@ -221,31 +223,67 @@ def _combine_bitfield(ints):
         return res
 
     @classmethod
-    def _find_detector_name(cls, data):
+    def _find_detector_names(cls, data):
+        # Find sources matching the pattern (raw or proc) for this detector type
+        raw_re  = re.compile(f'(?P<detname>{cls._det_name_pat}){cls._source_raw_pat}')
+        corr_re = re.compile(f'(?P<detname>{cls._det_name_pat}){cls._source_corr_pat}')
         detector_names = set()
         for source in data.instrument_sources:
-            m = cls._source_re.match(source)
-            if m:
-                detector_names.add(m.group('detname'))
+            if m := raw_re.match(source) or corr_re.match(source):
+                detector_names.add(m['detname'])
+        return detector_names
+
+    @classmethod
+    def _find_detector_name(cls, data):
+        detector_names = cls._find_detector_names(data)
+        # We want exactly 1 source
         if not detector_names:
-            raise SourceNameError(cls._source_re.pattern)
+            raise SourceNameError(f'{cls._det_name_pat}({cls._source_raw_pat}|{cls._source_corr_pat})')
         elif len(detector_names) > 1:
+            names_s = ', '.join(repr(n) for n in sorted(detector_names))
             raise ValueError(
-                "Multiple detectors found in the data: {}. "
-                "Pass a name to data.detector() to pick one.".format(
-                    ', '.join(repr(n) for n in detector_names)
-                )
+                f"Multiple detectors found in the data: {names_s}. "
+                f"Pass detector_name to {cls.__name__}() to pick one."
             )
         return detector_names.pop()
 
-    def _source_matches(self, data, detector_name):
+    @staticmethod
+    def _source_matches(data, pat):
+        source_re = re.compile(pat)
         for source in data.instrument_sources:
-            m = self._source_re.match(source)
-            if m and m.group('detname') == detector_name:
+            m = source_re.match(source)
+            if m:
                 yield source, int(m.group('modno'))
 
-    def _identify_sources(self, data, detector_name, modules=None):
-        source_to_modno = dict(self._source_matches(data, detector_name))
+    @classmethod
+    def _data_is_raw(cls, data, source: str):
+        # For most detectors, raw data is uint16 & corrected is float32.
+        # Overridden for AGIPD, where output dtype is configurable.
+        kd = data[source, cls._main_data_key]
+        return np.issubdtype(kd.dtype, np.integer)
+
+    @classmethod
+    def _identify_sources(cls, data, detector_name, modules=None, raw=None):
+        if raw is True:
+            pat = re.escape(detector_name) + cls._source_raw_pat
+            source_to_modno = dict(cls._source_matches(data, pat))
+            if not all(cls._data_is_raw(data, s) for s in source_to_modno):
+                # Older corrected data used the same names as raw
+                raise Exception("Raw data was requested, but only proc data was found")
```

Comment:
Nitpick, but why no `ValueError`?

Reply:
Good catch. I sometimes write Exception as a placeholder when I haven't figured out what the right type is.

In this scenario, what do you think about using `SourceNameError`? If we had had distinct source names from the beginning I think this would be a clear choice, but  it could be somewhat confusing if the relevant source *names* are there, just pointing to the wrong thing.

Reply:
Following discussion, I used `SourceNameError` when you want corrected data and only raw is there, because it can point to the new corrected source name, which will definitely be absent.

But when you want raw data and the `.../DET/...` source names point to corrected data, it seems confusing to say those sources are missing, so in that case I raise `ValueError`. Hopefully the disparity won't matter too much - the error message is probably the important thing in either case.